### ISSUE TITLE
RestClient can't perform EBF retry correctly.

### ISSUE THREAD

**Issue by thanatosjug** [Azure/autorest-clientruntime-for-java#410]
RestClient can't perform EBF retry, and do immediately retry when request fails with retryable exception.

Bug in RestClient.java in com.microsoft.rest.RestClient 

In line 432, it should be "this.retryStrategy = strategy".

         /**
         * Adds a retry strategy to the client.
         * @param strategy the retry strategy to add
         * @return the builder itself for chaining
         */
        public Builder withRetryStrategy(RetryStrategy strategy) {
            this.retryStrategy = retryStrategy;
            return this;
        }

**Comment by RikkiGibson** [comment]
A PR would be welcomed for this 👍

**Comment by RikkiGibson** [comment]
I fixed the obvious bug, but I looked into this a little more and it doesn't seem like RetryStrategy was ever fully implemented. Specifically, it doesn't look like there's any API that waits/notifies when the retry timeout has elapsed, or any usage of that API in RetryHandler.java. It just always retries immediately without waiting. The runtime v1 maintainers may wish to handle this one way or another.

If you need the delay behavior @thanatosjug you can consider implementing an [Interceptor](https://github.com/square/okhttp/wiki/Interceptors) in your own code that has the retry behavior you want and add it to your RestClient with `RestClient.withInterceptor(Interceptor)`.

Otherwise, if you want, you could also add some API to RetryStrategy for waiting/getting the time to wait from a RetryStrategy, modify RetryHandler to use it (I think Thread.sleep is fine), ideally with some test in RetryHandlerTests that checks a minimum elapsed time or something, and we'd be happy to take it and publish it. Otherwise it will go onto our backlog and I'm not sure when we'd be able to get to it.

Note that this stuff is changed around a bit in v2-- in there we have a RetryPolicy that performs a set number of retries immediately without waiting, and similarly you can implement your own RequestPolicy to do retry using our own HttpRequest/Response types and not the concrete types used by e.g. OkHttp.

**Comment by thanatosjug** [comment]
Yes, we also noticed that, and we understand it may takes time to implement and test this at your side.

We currently implemented our own retry policy and interceptor, and attached to our http client, and use reflection to disable the default retry strategy (can't set retry strategy because of the bug above). We are still looking for a fix for this from your team since it will be a clean solution.

**Comment by RikkiGibson** [comment]
@thanatosjug [client-runtime version 1.3.2 has been published](https://search.maven.org/#artifactdetails%7Ccom.microsoft.rest%7Cclient-runtime%7C1.3.2%7Cjar) to address the specific setter bug you pointed out. Fixes on RetryHandler itself may come in a future release. Thank you for pointing out the bug.
